spock-core/src/main/java/org/spockframework/mock/runtime/MockInteraction.java

```diff
@@ -65,18 +66,16 @@ public boolean matches(IMockInvocation invocation) {
   }
 
   @Override
-  public Object accept(IMockInvocation invocation) {
-    lock.writeLock().lock();
-    try {
-      acceptedInvocations.add(invocation);
-      if (acceptedInvocations.size() > maxCount) {
-        throw new TooManyInvocationsError(this, acceptedInvocations);
-      }
-
-      return responseGenerator == null ? null : responseGenerator.respond(invocation);
-    } finally {
-      lock.writeLock().unlock();
+  public Supplier<Object> accept(IMockInvocation invocation) {
+    acceptedInvocations.add(invocation);
```

Comment:
Is it Ok to access and modify `acceptedInvocations` without holding any lock?
So is a `MockInteraction` thread-local?

Reply:
It is not a thread-safe class, it is not thread-confined, and it is not safe to call `accept` without holding a lock.

The changes I made here are just rolling back the changes of the other PR, it was exactly like that before.

I marked the class as `@NotThreadSafe` explicitly now.
It could actually be made thread-safe lock-free simply by making the `acceptedInvocations` a `CopyOnWriteArraylist`.
Everything else is `final`, the `constraints` are effectively immutable, and all `responseGenerator`s are thread-safe now.

But the point is, that a `MockInteraction` instance is confined to a `MockController` and this `MockController` should only call methods on it while synchronizing on the mock controller to make sure the interaction matching and invocation counting is working properly, unlike with M2. So all accesses to `acceptedInvocations` actually **are** done under a lock, it just is external. So additionally making the class thread-safe could be done, but imho would just add unnecessary overhead, because it must anyway be used from synchronized blocks of `MockController` to ensure proper interaction matching and invocation counting.

The only thing where I'm not 100% sure is the `toString` method. This if called from a different thread could indeed show wrong values I think.